pravega/src/main/scala/akka/stream/alpakka/pravega/impl/PravegaReader.scala

```diff
@@ -0,0 +1,51 @@
+/*
+ * Copyright (C) 2016-2020 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.stream.alpakka.pravega.impl
+
+import akka.annotation.InternalApi
+import akka.stream.alpakka.pravega.ReaderSettings
+import akka.stream.stage.StageLogging
+import io.pravega.client.admin.ReaderGroupManager
+import io.pravega.client.stream.{ReaderGroup, ReaderGroupConfig, StreamCut, Stream => PravegaStream}
+
+@InternalApi private[pravega] trait PravegaReader extends PravegaCapabilities {
+  this: StageLogging =>
+
+  private lazy val readerGroupManager = ReaderGroupManager.withScope(scope, clientConfig)
+
+  def createReader[A](settings: ReaderSettings[A],
+                      streamName: String,
+                      start: StreamCut = StreamCut.UNBOUNDED,
+                      end: StreamCut = StreamCut.UNBOUNDED) = {
+    val readerGroup = createReaderGroup(settings, streamName)
+    val eventStreamReader = eventStreamClientFactory.createReader(
+      settings.readerId,
```

Comment:
One more novice question, can it happen that multiple `createReader` are invoked so that all the readers can consume events parallelly from the stream?
if yes, we need to ensure every reader under a `reader group` should have a unique id ...

Reply:
@shrids Good question.  Yes, it's possible.  In this API you get an Akka Streams source which encapsulates the Pravega Reader.

```scala
val source: Source[PravegaEvent[String], Future[Done]] = Pravega.source(scope, streamName)(readerSettings) // settings are passed implicitly in second parameter group
```

The stream hasn't run yet.  The `source` variable just contains a description of the stream.  You have to call its `run` method (or its variants) to _materialize_ the stream.  You can do this multiple times if you like, and each materialization will use the same parameters that were initially passed to `Pravega.source()`.

If the reader ID needs to be unique then we should add a GUID or enforce some other way of keeping it unique.  In Kafka consumer IDs are automatically generated for all consumers (unless otherwise overridden), so there are never two consumers with the same ID in the same consumer group.  Based on what you're saying, this isn't the case in Pravega?

Reply:
yes, at the moment the reader ids are not automatically generated.

```
 /**
     * Creates (or recreates) a new reader that is part of a {@link ReaderGroup}. The reader
     * will join the group and the members of the group will automatically rebalance among
     * themselves.
     * 
     * In the event that the reader dies, the method {@link ReaderGroup#readerOffline(String, Position)}
     * should be called, passing the last position of the reader (Usually done by storing the
     * position along with the output when it is processed.) Which will trigger redistribute the
     * events among the remaining readers.
     *
     * @param readerId A unique name (within the group) for this reader.
     * @param readerGroup The name of the group to join.
     * @param s The serializer for events.
     * @param config The reader's configuration.
     * @param <T> The type of events.
     * @return Newly created reader object that is a part of reader group
     */
    <T> EventStreamReader<T> createReader(String readerId, String readerGroup, Serializer<T> s, ReaderConfig config);
```

Reply:
Ok, then reader-id is now UUID randomly generated.